cylc/flow/tui/app.py

```diff
@@ -0,0 +1,606 @@
+#!/usr/bin/env python3
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) 2008-2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""The application control logic for Tui."""
+
+import sys
+
+import urwid
+from urwid import html_fragment
+
+from cylc.flow.exceptions import (
+    ClientError,
+    ClientTimeout
+)
+from cylc.flow.task_state import (
+    TASK_STATUSES_ORDERED,
+    TASK_STATUS_RUNAHEAD,
+    TASK_STATUS_WAITING,
+    TASK_STATUS_SUBMITTED,
+    TASK_STATUS_RUNNING,
+    TASK_STATUS_FAILED,
+)
+import cylc.flow.tui.overlay as overlay
+from cylc.flow.tui import (
+    BINDINGS,
+    FORE,
+    BACK,
+    JOB_COLOURS,
+    SUITE_COLOURS,
+)
+from cylc.flow.tui.tree import (
+    find_closest_focus,
+    translate_collapsing
+)
+from cylc.flow.tui.util import (
+    compute_tree,
+    dummy_flow,
+    # get_group_state,
+    get_job_icon,
+    get_task_icon,
+    get_task_status_summary,
+    get_workflow_status_str,
+    render_node
+)
+
+
+urwid.set_encoding('utf8')  # required for unicode task icons
+
+TREE_EXPAND_DEPTH = [2]
+
+QUERY = '''
+  query cli($taskStates: [String]){
+    workflows {
+      id
+      name
+      status
+      stateTotals
```

Comment:
Can also bring the cycle points, saving the work to have to iterate through the family proxies (I think the code is doing that in `tui.util`):

```graphql
cyclePoints: familyProxies (ids: ["root"]) {
  cyclePoint
}
```

Learned it from @dwsutherland in the CylcCon. But also not super important.

Reply:
Will leave this one until the delta work as the query will have to change then.